fastestimator/trace/io/csv_logger.py

```diff
@@ -60,39 +63,45 @@ def on_begin(self, data: Data) -> None:
         self.df_ins = pd.DataFrame(columns=base_keys)
 
     def on_epoch_end(self, data: Data) -> None:
-        tmpdic = {"mode": self.system.mode, "step": self.system.global_step, "epoch": self.system.epoch_idx}
-
-        if "*" in self.inputs:
-            for key, value in data.read_logs().items():
-                tmpdic[key] = self._parse_val(value)
-                if key not in self.df_agg.columns:
-                    self.df_agg[key] = ''
-        else:
-            for key in self.inputs:
-                if key == self.instance_id_key:
-                    continue
-                tmpdic[key] = self._parse_val(data[key])
-                if key not in self.df_agg.columns:
-                    self.df_agg[key] = ''
+        keys = set(self.inputs) if "*" not in self.inputs else set(self.inputs) | data.read_logs().keys()
+        keys = keys - {'*', self.instance_id_key}
+
+        tmpdic = {}
+        for key in keys:
+            tmpdic[key] = self._parse_val(data.read_logs().get(key, ''))
+            if key not in self.df_agg.columns:
+                self.df_agg[key] = ''
         for col in self.df_agg.columns:
-            if col not in tmpdic.keys():
+            if col not in tmpdic.keys() | {'mode', 'step', 'epoch'}:
```

Comment:
In this case the codacy proposal is invalid since older python versions don't support union of dictionary to set